hooks/command

```diff
@@ -13,64 +13,195 @@ fi
-    jobspec=$(echo "$jobspec" | jq ".spec.template.spec.$field=$buildkite_agent_value")
+    buildkite_agent_value="$(kubectl get pod "$(cat /etc/hostname)" -o json | jq ".spec.$field")"
+    job_spec="$(echo "$job_spec" | jq ".spec.template.spec.$field=$buildkite_agent_value")"
   done
 fi
 
-echo "$jobspec" | kubectl apply -f -
+if [[ "$print_resulting_job_spec" == "true" ]]; then
+  echo -e "Resulting k8s job spec:\n$job_spec"
+fi
+
+readonly job_apply_start_time="$SECONDS"
+job_apply_exit_code=""
+
+while [[ "$((SECONDS - job_apply_start_time))" -lt "$job_apply_timeout_sec" ]]
+do
+  set +e
+  echo "$job_spec" | kubectl apply -f -
+  job_apply_exit_code="$?"
+  set -e
+
+  if [[ "$job_apply_exit_code" == "0" ]]; then
+    break
+  else
+    echo "Attempt to apply the job failed, exit code '$job_apply_exit_code'"
+    sleep "$job_apply_retry_interval_sec"
+  fi
+done
+
+if [[ "$job_apply_exit_code" != "0" ]]; then
+  echo "Failed to apply job, exit code '$job_apply_exit_code'"
+  exit "$job_apply_exit_code"
+fi
 
 echo "Timeout: ${timeout}s"
 
-echo "+++ :kubernetes: Running image: ${BUILDKITE_PLUGIN_K8S_IMAGE}"
+echo "--- :kubernetes: Running image: ${BUILDKITE_PLUGIN_K8S_IMAGE}"
+
+counter="$timeout"
+while true
+do
+  set +e
+  pod_name=$(kubectl get pod -l "job-name=$job_name" --output=jsonpath="{.items[*].metadata.name}")
+  set -e
+
+  if [[ -n "$pod_name" ]]; then
+    break
+  else
+    sleep "$job_status_retry_interval_sec"
+    if [[ $timeout -gt 0 ]]; then
+      (( counter -= job_status_retry_interval_sec ))
+    fi
+  fi
+done
+
+echo "Pod is running: $pod_name"
+
+echo "--- :kubernetes: bootstrap container"
+tail_logs "$pod_name" "bootstrap" "$bootstrap_container_log_complete_marker_file"
 
-tail_logs &
+echo "+++ :kubernetes: step container"
+tail_logs "$pod_name" "step" "$step_container_log_complete_marker_file" &
 
-sleeper=2
-counter=${timeout}
+counter="$timeout"
 jobstatus=""
 while [[ -z "$jobstatus" ]] ; do
+  set +e
   jobstatus=$(kubectl get job "${job_name}" -o 'jsonpath={.status.conditions[].type}')
-  sleep $sleeper
+  set -e
+
+  if [[ -n "$jobstatus" ]]; then
+    break
+  else
+    sleep "$job_status_retry_interval_sec"
+  fi
+
   if [[ $timeout -gt 0 ]]; then
-    (( counter -= sleeper )) || jobstatus="timeout"
+    (( counter -= job_status_retry_interval_sec )) || jobstatus="timeout"
   fi
 done
 
-echo
 echo "--- :kubernetes: Job status: $jobstatus"
-status=1
+
+# Wait for logs to be fully printed, printing runs in a separate process and we're racing with it.
+readonly log_complete_start_time="$SECONDS"
+while [[ "$(cat "$step_container_log_complete_marker_file")" != "0" ]] && [[ "$((SECONDS - log_complete_start_time))" -lt "$log_complete_timeout_sec" ]]
+do
+  sleep "$log_complete_retry_interval_sec"
+done
+
+status=""
 if [[ "$jobstatus" == "Complete" ]] ; then
   echo "success"
-  status=0
+  status="0"
+else
+  while true
```

Comment:
This looks like we'll keep looping if we keep failing to get the init container status? Let's break out of the loop after a hardcoded number of attempts or timeout instead?